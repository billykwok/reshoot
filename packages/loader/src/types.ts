import { type CoreImageMeta, type CoreOptions } from '@reshoot/types';

export type Field =
  | string
  | number
  | boolean
  | ((...args: unknown[]) => unknown)
  | { [key: string]: Field }
  | Field[];

export type LoaderImageMeta = CoreImageMeta & { src: string };

export type LoaderOptions = {
  meta: CoreOptions;
  shape: (output: LoaderImageMeta, resourcePath: string) => Field;
  maxWidth: number;
  filename: string;
  outputPath: string | ((path: string) => string);
  publicPath: string | ((filename: string) => string);
  emitFile: boolean;
  esModule: boolean;
};
