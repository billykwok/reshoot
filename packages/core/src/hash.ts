import crypto from 'node:crypto';
import { type CoreOptions } from '@reshoot/types';

export function generateHash(content: Buffer, options: CoreOptions): string {
  return crypto
    .createHash('md5')
    .update(JSON.stringify(options))
    .update(content)
    .digest('hex');
}
