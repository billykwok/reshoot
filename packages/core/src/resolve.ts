import color from 'color';
import {
  AspectRatioFormat,
  AspectRatioType,
  type CoreOptions,
} from '@reshoot/types';
import { type Sharp } from 'sharp';

export function resolveAspectRatio(
  { width, height }: { width?: number; height?: number },
  {
    aspectRatioType,
    aspectRatioFormat,
    aspectRatioDecimal,
  }: Pick<
    CoreOptions,
    'aspectRatioType' | 'aspectRatioFormat' | 'aspectRatioDecimal'
  >
): number {
  const scale = aspectRatioFormat === AspectRatioFormat.Percent ? 100 : 1;
  const rounder = 10 ** aspectRatioDecimal;
  switch (aspectRatioType) {
    case AspectRatioType.WidthByHeight:
      return Math.floor((width / height) * scale * rounder) / rounder;
    case AspectRatioType.HeightByWidth:
      return Math.floor((height / width) * scale * rounder) / rounder;
  }
}

export async function resolvePlaceholder(
  image: Sharp,
  {
    placeholderSize,
    placeholderQuality,
  }: Pick<CoreOptions, 'placeholderSize' | 'placeholderQuality'>
): Promise<string> {
  return (
    'data:image/png;base64,' +
    (
      await image
        .resize({
          fit: 'inside',
          width: placeholderSize,
          height: placeholderSize,
          withoutEnlargement: true,
        })
        .png({
          compressionLevel: 9,
          quality: placeholderQuality,
          palette: true,
          colors: 16,
        })
        .toBuffer()
    ).toString('base64')
  );
}

export async function resolveColor(
  image: Sharp,
  colorOptions: string
): Promise<string> {
  return colorOptions
    ? colorOptions === 'transparent'
      ? colorOptions
      : colorOptions.toLowerCase()
    : color((await image.stats()).dominant)
        .hex()
        .toLocaleLowerCase();
}
