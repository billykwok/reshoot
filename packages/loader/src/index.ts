import { loader } from 'webpack';

import resolveOptions from './resolveOptions';
import extractPassThroughProperties from './extractPassThroughProperties';
import resolveMimeAndExt from './resolveMimeAndExt';
import toScriptString from './toScriptString';
import resolveColor from './resolveColor';
import createSharp from './createSharp';
import createDataUrl from './createDataUrl';
import emit from './emit';
import resolveAspectRatio from './resolveAspectRatio';
import resize from './resize';
import createHash from './createHash';
import cache from './cache';
import resolveDefaultOptions from './defaultOptions';
import createOutputPathResolver from './createOutputPathResolver';

async function reshootLoader(this: loader.LoaderContext, content: string) {
  this.cacheable(true);
  const callback = this.async();
  const defaultOptions = resolveDefaultOptions(this.mode);
  const options = resolveOptions(this, defaultOptions);
  const context = options.context || this.rootContext;
  const output = extractPassThroughProperties(options, defaultOptions);

  const { shape: outputShape } = options;

  const resolveOutputPath = createOutputPathResolver(options);
  const image = createSharp(this.resourcePath);
  const hash = createHash(await image.content(), options);

  const cachedOutput = await cache.invalidateCache(hash);
  if (cachedOutput) {
    if (options.emitFile && cachedOutput.files.length > 0) {
      const emitCache = async ({ outputPath, cachePath }) => {
        const content = await cache.readCacheFile(cachePath);
        this.emitFile(resolveOutputPath(outputPath), content, null);
      };
      await Promise.all(cachedOutput.files.map(emitCache));
    }
    image.close();
    return callback(null, cachedOutput.output);
  }

  const [saver, metadata] = await Promise.all([
    cache.createSaver(hash),
    image.metadata()
  ]);
  const [mime, ext] = resolveMimeAndExt(this, options.forceFormat);
  const rawPath = emit(
    this,
    context,
    content,
    hash,
    metadata.width,
    ext,
    options,
    resolveOutputPath,
    saver
  );

  output.src = rawPath;

  if (outputShape.mime) {
    output.mime = mime;
  }

  if (outputShape.aspectRatio) {
    output.aspectRatio = resolveAspectRatio(metadata, options.aspectRatio);
  }

  if (outputShape.color) {
    output.color = await resolveColor(image, options.color, options.disable);
  }

  if (options.disable) {
    if (outputShape.placeholder) {
      output.placeholder = rawPath;
    }
    if (outputShape.srcSet) {
      output.srcSet = null;
    }
    const serializedOutput = toScriptString(outputShape, output);
    saver.save(serializedOutput);
    image.close();
    return callback(null, serializedOutput);
  }

  const placeholder = options.placeholder;
  const [promises, widths] = resize(
    image,
    metadata,
    placeholder,
    mime,
    options
  );
  if (widths.size === 0) {
    if (outputShape.placeholder) {
      output.placeholder = null;
    }
    if (outputShape.srcSet) {
      output.srcSet = null;
    }
    const serializedOutput = toScriptString(outputShape, output);
    saver.save(serializedOutput);
    image.close();
    return callback(null, serializedOutput);
  }

  const [placeholderData, ...imagesData] = await Promise.all(promises);

  if (outputShape.placeholder) {
    output.placeholder = createDataUrl(
      mime,
      placeholderData.content,
      placeholder.trimDataUrl
    );
  }

  if (outputShape.srcSet) {
    const paths = new Map<number, string>();
    imagesData.forEach(({ content, width }) =>
      paths.set(
        width,
        emit(
          this,
          context,
          content,
          hash,
          width,
          ext,
          options,
          resolveOutputPath,
          saver
        )
      )
    );
    if (paths.size > 0) {
      output.srcSet = options.srcSet
        .filter(size => paths.has(size))
        .map(size => `${paths.get(widths.get(size))} ${size}w`);
    } else {
      output.srcSet = null;
    }
  }

  const serializedOutput = toScriptString(outputShape, output);
  saver.save(serializedOutput);
  image.close();
  return callback(null, serializedOutput);
}

export const raw = true;

export default reshootLoader;
