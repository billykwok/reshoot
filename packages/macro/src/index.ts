import reshootMacro from './macro';
export default reshootMacro;
