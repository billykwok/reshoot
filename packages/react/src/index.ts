import { memo, useRef, useImperativeHandle, forwardRef } from 'react';
import { css } from 'linaria';
import assign from 'object-assign';
import type { RefObject, SyntheticEvent } from 'react';

import h from './h';
import Placeholder from './Placeholder';
import Img from './Img';
import Message from './Message';
import State from './state';
import useImage from './useImage';
import useCx from './useCx';

type Props = {
  src: string;
  alt: string;
  aspectRatio: number;
  className?: string;
  blur?: number;
  color?: string;
  placeholder?: string;
  srcSet?: string;
  target?: string;
  href?: string;
  messages?: {
    [State.MANUAL]: string;
    [State.OFFLINE]: string;
    [State.ERROR]: string;
  };
  onClick?: (e: SyntheticEvent<Element, MouseEvent>) => void;
};

const asContainer = css`
  overflow: hidden;
  position: relative;
  text-decoration: none;
  padding: 0;
  border: none;
  outline: none;
`;

const asButtonContainer = css`
  cursor: pointer;
`;

const Reshoot = (
  {
    src,
    className,
    alt,
    aspectRatio,
    blur = 20,
    color = 'transparent',
    placeholder = null,
    srcSet = null,
    target = '_self',
    href = null,
    messages = {
      MANUAL: 'Not autoloaded in slow network',
      OFFLINE: 'Browser is offline',
      ERROR: 'Fail to load',
    },
    onClick,
    ...rest
  }: Props,
  ref: RefObject<HTMLElement>
) => {
  const innerRef = useRef<HTMLElement>(null);
  useImperativeHandle(ref, () => innerRef.current);
  const [state, setState, download] = useImage(innerRef, src, srcSet);
  const cx = useCx(className);

  const children = [
    h(Placeholder, { color, aspectRatio }),
    h(Img, { color, placeholder, src, srcSet, alt, state, blur }),
    h(Message, { state, text: messages[state] as string }),
  ];

  if (state !== State.INITIAL && state !== State.LOADED) {
    return h(
      'button',
      assign(
        {
          ref: innerRef,
          className: cx(asContainer, asButtonContainer),
          onClick: (e: SyntheticEvent<Element, MouseEvent>) => {
            e.stopPropagation();
            e.preventDefault();
            setState(() => State.INITIAL);
            download();
          },
        },
        rest
      ),
      ...children
    );
  }
  if (href) {
    return h(
      'a',
      assign(
        { ref: innerRef, className: cx(asContainer), target, href, onClick },
        rest
      ),
      ...children
    );
  }
  return h(
    'div',
    assign({ ref: innerRef, className: cx(asContainer) }, rest),
    ...children
  );
};

export default memo(forwardRef(Reshoot));
