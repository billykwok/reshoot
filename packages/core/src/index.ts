import sharp from 'sharp';
import {
  AspectRatioFormat,
  AspectRatioType,
  type CoreOptions,
  type CoreImageMeta,
} from '@reshoot/types';
import { generateHash } from './hash';
import {
  resolveAspectRatio,
  resolveColor,
  resolvePlaceholder,
} from './resolve';

export const DEFAULT_OPTIONS: CoreOptions = Object.freeze({
  color: null,
  maxWidth: Infinity,
  placeholderSize: 8,
  placeholderQuality: 10,
  aspectRatioType: AspectRatioType.HeightByWidth,
  aspectRatioFormat: AspectRatioFormat.Percent,
  aspectRatioDecimal: 4,
});

export async function extractMeta(
  content: Buffer,
  options: Partial<CoreOptions> & Record<string, any> = {}
): Promise<CoreImageMeta> {
  const mergedOptions: CoreOptions = { ...DEFAULT_OPTIONS, ...options };

  const image = sharp(content);
  const [metadata, placeholder, color] = await Promise.all([
    image.metadata(),
    resolvePlaceholder(image, mergedOptions),
    resolveColor(image, options.color),
  ]);

  return {
    hash: generateHash(content, mergedOptions),
    width: metadata.width,
    height: metadata.height,
    aspectRatio: resolveAspectRatio(metadata, mergedOptions),
    placeholder,
    color,
  };
}

export {
  AspectRatioFormat,
  AspectRatioType,
  type CoreImageMeta,
  type CoreOptions,
} from '@reshoot/types';
