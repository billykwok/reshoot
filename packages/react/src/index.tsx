import {
  type CSSProperties,
  type ForwardedRef,
  type MutableRefObject,
  type MouseEvent,
  type ReactNode,
  type Ref,
  type RefCallback,
  forwardRef,
  useCallback,
  useRef,
  useState,
} from 'react';
import { css } from '@linaria/core';
import useLifecycleRef from 'use-lifecycle-ref';
import type { ImageMeta } from '@reshoot/types';

import useLazyImage from './useLazyImage';

declare module 'react' {
  interface CSSProperties {
    [index: `--${string}`]: any;
  }
}

const enum State {
  INITIAL,
  FADING,
  LOADED,
  ERROR,
}

const container = css`
  position: relative;
  overflow: hidden;
  background: var(--c);
  &:before {
    content: '';
    display: block;
    width: 100%;
    height: 0;
    padding: 0 0 var(--r);
  }
  > * {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }
  > div[aria-hidden='true'] {
    background: var(--c);
    > img {
      width: 100%;
      height: 100%;
      filter: blur(2rem);
    }
  }
`;

const error = css`
  display: flex;
  backdrop-filter: blur(0.5rem);
  align-items: center;
  justify-content: center;
  flex-direction: column;
  color: #f7f7f7;
  text-shadow: #000 0 0 1rem;
  background: rgba(0, 0, 0, 0.5);
  > p {
    margin: 0 0 0.5rem;
  }
  > p:first-child {
    font-size: 1.125rem;
  }
`;

const fadeOut = css`
  animation: fadeOut 0.75s 1 forwards;
  @keyframes fadeOut {
    0% {
      opacity: 1;
    }
    100% {
      opacity: 0;
    }
  }
`;

export type Props = {
  meta: ImageMeta;
  container?:
    | string
    | ((props: {
        ref: Ref<HTMLElement>;
        className: string;
        children: ReactNode;
        style: CSSProperties;
      }) => ReactNode);
  ref?: RefCallback<HTMLElement> | MutableRefObject<HTMLElement>;
  imgRef?: RefCallback<HTMLImageElement> | MutableRefObject<HTMLImageElement>;
  alt?: string;
  className?: string;
  style?: Record<string, any>;
  onClick?: (e: MouseEvent<HTMLElement>) => void;
  [key: string]: any;
};

const Img = forwardRef(function RefReceivingImg(
  {
    meta,
    container: Container = 'div',
    imgRef,
    alt,
    className,
    style,
    ...props
  }: Props,
  ref: ForwardedRef<HTMLElement>
): JSX.Element {
  const { src, srcSet, sizes, aspectRatio, placeholder, color } = meta;
  const [state, setState] = useState(State.INITIAL);
  const instantFade = useRef(false);
  const onLoad = useCallback(
    () => setState(instantFade.current ? State.LOADED : State.FADING),
    []
  );
  const onError = useCallback(() => setState(State.ERROR), []);
  useLazyImage(src, srcSet, sizes, onLoad, onError);
  const _imgRef = useLifecycleRef<HTMLImageElement>({
    onAttach: (element) => {
      if (element.complete) {
        instantFade.current = true;
      }
    },
    ref: imgRef,
  });
  return (
    <Container
      ref={ref}
      className={container + (className ? ' ' + className : '')}
      style={{ '--c': color, '--r': `${aspectRatio}%`, ...style }}
      {...props}
    >
      <img
        ref={_imgRef}
        src={src}
        srcSet={srcSet}
        sizes={sizes}
        alt={alt}
        crossOrigin=""
        aria-busy={State.LOADED !== state && State.ERROR !== state}
      />
      {State.LOADED !== state && (
        <div
          className={
            'js-only' +
            (State.FADING === state && !instantFade.current
              ? ' ' + fadeOut
              : '')
          }
          onAnimationEnd={
            State.FADING === state ? () => setState(State.LOADED) : null
          }
          aria-hidden="true"
        >
          <img src={placeholder} alt="" />
        </div>
      )}
      {State.ERROR === state && (
        <div className={error}>
          <p>Failed to load image</p>
          <p>Please try refreshing the page</p>
        </div>
      )}
    </Container>
  );
});

export { type ImageMeta } from '@reshoot/types';

export default Img;
