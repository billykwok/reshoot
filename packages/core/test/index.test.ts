import path from 'node:path';
import { AspectRatioFormat, AspectRatioType } from '@reshoot/types';
import { describe, expect, test } from '@jest/globals';
import { fileURLToPath } from 'node:url';
import { readFile } from 'node:fs/promises';

import { DEFAULT_OPTIONS, extractMeta } from '../src';

const __dirname = fileURLToPath(new URL('.', import.meta.url));

const IMAGE_PATH = path.resolve(__dirname, '../../../__fixtures__/lena.png');
const SNAPSHOT_MATCHER = {
  placeholder: expect.stringMatching(
    /data:image\/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA/
  ),
};

describe('defaultOptions', () => {
  test('basic test', () => {
    expect(DEFAULT_OPTIONS).toMatchSnapshot();
  });
});

describe('extractMeta', () => {
  test('should extract meta without options', async () => {
    const metaData = await extractMeta(await readFile(IMAGE_PATH));
    expect(metaData).toMatchSnapshot(SNAPSHOT_MATCHER);
  });

  test('should extract meta with color transparent', async () => {
    const metaData = await extractMeta(await readFile(IMAGE_PATH), {
      color: 'transparent',
    });
    expect(metaData).toMatchSnapshot(SNAPSHOT_MATCHER);
  });

  test('should extract meta with color override', async () => {
    const metaData = await extractMeta(await readFile(IMAGE_PATH), {
      color: '#eee',
    });
    expect(metaData).toMatchSnapshot(SNAPSHOT_MATCHER);
  });

  test('should extract meta with aspectRatioFormat `ratio`', async () => {
    const metaData = await extractMeta(await readFile(IMAGE_PATH), {
      aspectRatioFormat: AspectRatioFormat.Ratio,
    });
    expect(metaData).toMatchSnapshot(SNAPSHOT_MATCHER);
  });

  test('should extract meta with aspectRatioType `widthByHeight`', async () => {
    const metaData = await extractMeta(await readFile(IMAGE_PATH), {
      aspectRatioType: AspectRatioType.WidthByHeight,
    });
    expect(metaData).toMatchSnapshot(SNAPSHOT_MATCHER);
  });
});
