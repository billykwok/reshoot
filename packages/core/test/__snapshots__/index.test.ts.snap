// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultOptions basic test 1`] = `
Object {
  "aspectRatioDecimal": 4,
  "aspectRatioFormat": "percent",
  "aspectRatioType": "heightByWidth",
  "color": null,
  "maxWidth": Infinity,
  "placeholderQuality": 10,
  "placeholderSize": 8,
}
`;

exports[`extractMeta should extract meta with aspectRatioFormat \`ratio\` 1`] = `
Object {
  "aspectRatio": 1,
  "color": "#c85858",
  "hash": "7fb419d29095e35c0360352224eea0f9",
  "height": 512,
  "placeholder": StringMatching /data:image\\\\/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA/,
  "width": 512,
}
`;

exports[`extractMeta should extract meta with aspectRatioType \`widthByHeight\` 1`] = `
Object {
  "aspectRatio": 100,
  "color": "#c85858",
  "hash": "5bc91dcfcabb72b70db3cc31ee20a935",
  "height": 512,
  "placeholder": StringMatching /data:image\\\\/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA/,
  "width": 512,
}
`;

exports[`extractMeta should extract meta with color override 1`] = `
Object {
  "aspectRatio": 100,
  "color": "#eee",
  "hash": "affff220bfa0e17de888b3de33fc4023",
  "height": 512,
  "placeholder": StringMatching /data:image\\\\/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA/,
  "width": 512,
}
`;

exports[`extractMeta should extract meta with color transparent 1`] = `
Object {
  "aspectRatio": 100,
  "color": "transparent",
  "hash": "c2bcac89cb8fa278b45f0b3f74aa187c",
  "height": 512,
  "placeholder": StringMatching /data:image\\\\/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA/,
  "width": 512,
}
`;

exports[`extractMeta should extract meta without options 1`] = `
Object {
  "aspectRatio": 100,
  "color": "#c85858",
  "hash": "5f932a17e3ea4c07f6d22128ec9ff65b",
  "height": 512,
  "placeholder": StringMatching /data:image\\\\/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA/,
  "width": 512,
}
`;
